File: snippets/dart_snippets/lib/buy_btc.dart
import 'package:dart_snippets/sdk_instance.dart';
import 'package:flutter_breez_liquid/flutter_breez_liquid.dart';

Future<OnchainPaymentLimitsResponse> fetchOnchainLimits() async {
  // ANCHOR: onchain-limits
  OnchainPaymentLimitsResponse currentLimits =
      await breezSDKLiquid.instance!.fetchOnchainLimits();
  print("Minimum amount: ${currentLimits.receive.minSat} sats");
  print("Maximum amount: ${currentLimits.receive.maxSat} sats");
  // ANCHOR_END: onchain-limits
  print(currentLimits);
  return currentLimits;
}

Future<PrepareBuyBitcoinResponse> prepareBuyBitcoin(
    OnchainPaymentLimitsResponse currentLimits) async {
  // ANCHOR: prepare-buy-btc
  PrepareBuyBitcoinRequest req = PrepareBuyBitcoinRequest(
      provider: BuyBitcoinProvider.moonpay,
      amountSat: currentLimits.receive.minSat);
  PrepareBuyBitcoinResponse prepareRes =
      await breezSDKLiquid.instance!.prepareBuyBitcoin(req: req);

  // Check the fees are acceptable before proceeding
  BigInt receiveFeesSat = prepareRes.feesSat;
  print("Fees: ${receiveFeesSat} sats");
  // ANCHOR_END: prepare-buy-btc
  return prepareRes;
}

Future<String> buyBitcoin(PrepareBuyBitcoinResponse prepareResponse) async {
  // ANCHOR: buy-btc
  BuyBitcoinRequest req = BuyBitcoinRequest(prepareResponse: prepareResponse);
  String url = await breezSDKLiquid.instance!.buyBitcoin(req: req);
  // ANCHOR_END: buy-btc
  return url;
}

File: snippets/swift/BreezSDKExamples/Sources/BuyBtc.swift
import BreezSDKLiquid
import Foundation

func fetchOnchainLimits(sdk: BindingLiquidSdk) {
    //  ANCHOR: onchain-limits
    if let currentLimits = try? sdk.fetchOnchainLimits() {
        print("Minimum amount, in sats: \(currentLimits.receive.minSat)")
        print("Maximum amount, in sats: \(currentLimits.receive.maxSat)")
    }
    // ANCHOR_END: onchain-limits
}

func prepareBuyBitcoin(sdk: BindingLiquidSdk, currentLimits: OnchainPaymentLimitsResponse) {
    // ANCHOR: prepare-buy-btc
    let req = PrepareBuyBitcoinRequest(
        provider: .moonpay,
        amountSat: currentLimits.receive.minSat)

    if let prepareResponse = try? sdk.prepareBuyBitcoin(req: req) {
        // Check the fees are acceptable before proceeding
        let receiveFeesSat = prepareResponse.feesSat;
        print("Fees: \(receiveFeesSat) sats")
    }
    // ANCHOR_END: prepare-buy-btc
}

func buyBitcoin(sdk: BindingLiquidSdk, prepareResponse: PrepareBuyBitcoinResponse) -> String? {
    // ANCHOR: buy-btc
    let req = BuyBitcoinRequest(prepareResponse: prepareResponse)
    let url = try? sdk.buyBitcoin(req: req)
    // ANCHOR_END: buy-btc
    return url
}
